pkg/sdn/plugin/sdn-cni-plugin/openshift-sdn.go

```diff
@@ -0,0 +1,511 @@
+		err         error
+		podIP       string
+	)
+
+	containerNs, err := ns.GetNS(netns)
+	if err != nil {
+		return nil, "", "", fmt.Errorf("failed to get container netns: %v", err)
+	}
+	defer containerNs.Close()
+
+	err = containerNs.Do(func(ns.NetNS) error {
+		contVeth, err = netlink.LinkByName(containerIfname)
+		if err != nil {
+			return err
+		}
+		peerIfindex = contVeth.Attrs().ParentIndex
+
+		addrs, err := netlink.AddrList(contVeth, syscall.AF_INET)
+		if err != nil {
+			return fmt.Errorf("failed to get container IP addresses: %v", err)
+		}
+		if len(addrs) == 0 {
+			return fmt.Errorf("container had no addresses")
+		}
+		podIP = addrs[0].IP.String()
+
+		return nil
+	})
+	if err != nil {
+		return nil, "", "", fmt.Errorf("failed to inspect container interface: %v", err)
+	}
+
+	hostVeth, err := netlink.LinkByIndex(peerIfindex)
+	if err != nil {
+		return nil, "", "", fmt.Errorf("failed to get host veth: %v", err)
+	}
+
+	return hostVeth, contVeth.Attrs().HardwareAddr.String(), podIP, nil
+}
+
+func addMacvlan(netns string) error {
+	var defIface netlink.Link
+	var err error
+
+	// Find interface with the default route
+	routes, err := netlink.RouteList(nil, netlink.FAMILY_V4)
+	if err != nil {
+		return fmt.Errorf("failed to read routes: %v", err)
+	}
+
+	for _, r := range routes {
+		if r.Dst == nil {
+			defIface, err = netlink.LinkByIndex(r.LinkIndex)
+			if err != nil {
+				return fmt.Errorf("failed to get default route interface: %v", err)
+			}
+		}
+	}
+	if defIface == nil {
+		return fmt.Errorf("failed to find default route interface")
+	}
+
+	containerNs, err := ns.GetNS(netns)
+	if err != nil {
+		return fmt.Errorf("failed to get container netns: %v", err)
+	}
+	defer containerNs.Close()
+
+	return containerNs.Do(func(ns.NetNS) error {
+		err := netlink.LinkAdd(&netlink.Macvlan{
+			LinkAttrs: netlink.LinkAttrs{
+				MTU:         defIface.Attrs().MTU,
+				Name:        "macvlan0",
+				ParentIndex: defIface.Attrs().Index,
+			},
+			Mode: netlink.MACVLAN_MODE_PRIVATE,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to create macvlan interface: %v", err)
+		}
+		l, err := netlink.LinkByName("macvlan0")
+		if err != nil {
+			return fmt.Errorf("failed to create find macvlan interface: %v", err)
+		}
+		err = netlink.LinkSetUp(l)
+		if err != nil {
+			return fmt.Errorf("failed to set macvlan interface up: %v", err)
+		}
+		return nil
+	})
+}
+
+func getIPAMConfig(podConfig *PodConfig) ([]byte, error) {
+	nodeNet, err := types.ParseCIDR(podConfig.nodeNetwork)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing node network '%s': %v", podConfig.nodeNetwork, err)
+	}
+	clusterNet, err := types.ParseCIDR(podConfig.clusterNetwork)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing cluster network '%s': %v", podConfig.clusterNetwork, err)
+	}
+
+	type hostLocalIPAM struct {
+		Type   string        `json:"type,omitempty"`
+		Subnet types.IPNet   `json:"subnet"`
+		Routes []types.Route `json:"routes"`
```

Comment:
is there any reason to specify initial-caps field names with all-lowercase json overrides, rather than just specifying all-lowercase field names? (And you never omit any of the fields, so the "omitempty"s are also not needed)